harvester_e2e_tests/integration/test_upgrade.py

```diff
@@ -1,121 +1,1231 @@
+    if req.status_code == 200:
+        checksum = req.text.splitlines()[0].split()[0]
+
+    return api_client.versions.create(version, isoURL, checksum)
+
+
+def _get_all_nodes(api_client):
+    code, data = api_client.hosts.get()
+    assert code == 200, (
+       f"Failed to get nodes: {code}, {data}")
+    return data['data']
 
 
-@pytest.fixture(scope="module")
-def cluster_state(self):
+def _get_master_and_worker_nodes(api_client):
+    nodes = _get_all_nodes(api_client)
+    master_nodes = []
+    worker_nodes = []
+    for node in nodes:
+        if node['metadata']['labels'].get(CONTROL_PLANE_LABEL) == "true":
+            master_nodes.append(node)
+        else:
+            worker_nodes.append(node)
+    return master_nodes, worker_nodes
+
+
+def _is_installed_version(api_client, target_version):
+    code, data = api_client.upgrades.get()
+    if code == 200 and len(data['items']) > 0:
+        for upgrade in data['items']:
+            if upgrade.get('spec', {}).get('version') == target_version and \
+               upgrade['metadata']['labels'].get(UPGRADE_STATE_LABEL) == 'Succeeded':
+                return True
+    return False
+
+
+@pytest.fixture(scope="class")
+def network(request, api_client, cluster_state):
+    code, data = api_client.networks.get()
+    assert code == 200, (
+        "Failed to get networks: %s" % (data))
+
+    vlan_id = request.config.getoption('--vlan-id') or 1
+    for network in data['items']:
+        if network['metadata']['labels'].get(NETWORK_VLAN_ID_LABEL) == f"{vlan_id}":
+            cluster_state.network = network
+            return
+
+    raise AssertionError("Failed to find a routable vlan network")
+
+
+@pytest.fixture(scope="class")
+def unique_id(cluster_state, unique_name):
+    cluster_state.unique_id = unique_name
+
+
+@pytest.fixture(scope="class")
+def cluster_state():
     class ClusterState:
+        vm1 = None
+        vm2 = None
+        vm3 = None
         pass
 
     return ClusterState()
 
 
-@pytest.fixture(scope="module")
-def cluster_prereq(self, api_client, cluster_state):
-    # create cluster network
-    # create a routable vlan
-    cluster_state.clusternetwork = "cluster network name"
-    cluster_state.vm_network = ("name", "vlanid", "cluster network name")
-    cluster_state.vlan = "routable vlan id"
+@pytest.fixture(scope="class")
+def cluster_prereq(request, cluster_state, prepare_dependence, unique_id, network, base_sc,
+                   ubuntu_image, new_sc):
+    assert request.config.getoption('--upgrade-iso-url'), (
+        "upgrade-iso-url must be not empty")
+
+    assert request.config.getoption('--upgrade-iso-checksum'), (
+        "upgrade-iso-checksum must be not empty")
+
+    if request.config.getoption('--upgrade-target-version'):
+        cluster_state.version_verify = True
+        cluster_state.version = request.config.getoption('--upgrade-target-version')
+    else:
+        cluster_state.version = f"version-{cluster_state.unique_id}"
```

Comment:
`cluster_state.version_verify` need to be assigned `False` when there is no option configured for `target-version`.
otherwise we can not use `if cluster_state.version_verify` in following test case.